### DISCUSSION TITLE
301 to 404 error | SSR on Client Side Navigation

### DISCUSSION THREAD

**Discussion by binoiii** [vercel/next.js#36796] [Help]
The JSON file is fetched successfully in the development environment when I do client-side navigation on an SSR rendered page. 
![image](https://user-images.githubusercontent.com/54553660/167574968-31d0ac0b-76df-4f2e-a448-a0be53087478.png)

On the other hand, in production build, it triggers 301 which will result in 404. 
![image](https://user-images.githubusercontent.com/54553660/167575632-c31fb560-8781-4b1d-8126-e3fde344988d.png)
![image](https://user-images.githubusercontent.com/54553660/167575750-9aee3ae1-b5a1-469f-abce-2b098743a813.png)

**Comment by icyJoseph** [comment]
Got a repo to take a look at?

I reckon that you have a branch in your `getServerSideProps`(GSSP) that does redirect? and that's redirecting the page to a destination that doesn't exist?

If there's no repo to look at, could you paste pseudo-code of your GSSP? 

> Make sure that your env vars, and other configuration are also in place, when building for production.

**Reply by binoiii** [reply]
Here is the code snippet of my GSSP. 

`export const getServerSideProps = async ({ req }) => {
  console.log(req, "req")
  return { props: { message: "test message v1" } }
}`

Btw, thanks @icyJoseph. Appreciate the help.

**Reply by icyJoseph** [reply]
No worries.

Alright, I see no issues on a vanilla project. I go from `/a` to `/b`, using `next/link`, where `/b` is using that `GSSP` and no issues show in production.

Are you using redirects, middleware, rewrites, or have anything else that could be interfering with the request?

I can't think of anything else that could be triggering this.

In the snippet you shared, I did change:

```js
export const getServerSideProps = async ({
  req,
}) => {
  console.log(req, "req"); // added a ; here
  return { props: { message: "test message v1" } };
};
```

**Comment by binoiii** [comment]
Here is my test component and I think nothing has been done to interfere with the request. Tried both for Link and router to navigate to the page but got the same result in the production build. 

Yeah I wonder why the _next/app/.../.json can't be found.

Btw, the json file is built during request time right?

```
import React from "react"
import Link from "next/link"
import { useRouter } from "next/router"

const Test = () => {
  const router = useRouter()

  const handleClick = (event) => {
    event.preventDefault()
    router.push("/search-jobs")
  }

  return (
    <>
      <button onClick={ handleClick }>click</button>
      <Link href="/search-jobs">
        <a>link</a>
      </Link>
    </>
  )
}
export default Test
```

**Reply by icyJoseph** [reply]
Yeah, nope, it all works for me.

Yes, if there's no explicit caching being set on the GSSP response, it is generated on every request (on-demand).

Could it be some kind of problem with `http` and `https`? 

If you have a repository with this, you could invite me and I can take a look, otherwise, this might take some time to debug.

**Reply by binoiii** [reply]
@icyJoseph Yeah as much as I want you to invite and have a look so we can debug faster. However, the repo is private.
The vanilla code that I shared is just a simplified version. 

>Could it be some kind of problem with http and https?

Any thoughts where I could start checking regards to this problem?

Also, is there a chance or a way that we could check the contents of /_next/data folder?

**Reply by icyJoseph** [reply]
Without looking at a repo, or a more approximate version of the code, it is really hard to tell.

The protocol (http/https), just came to mind, but if you wonder where to start checking, then it means, that most likely, it isn't the problem.

Not sure what you'd expect to find in that folder.  

One thing you could do is print something at the top of `GSSP`, and make sure it still logs when you run in production mode. That should start to give us some hints.
